### DISCUSSION TITLE
Kafka Test Companion feedback and improvements

### DISCUSSION THREAD

**Discussion by loicmathieu** [smallrye/smallrye-reactive-messaging#1852] [Ideas]
I recently migrated a bunch of integration tests from raw Kafka consumer/producer API to the Kafka Companion.
The API is very convenient but here a some feedback and possible improvements.

1. There is shortcut methods for consuming / producing strings and primitive but as soon as you want to use custom serializers and deserializers you're ending up to write a lot of `companion.consumeWithDeserializer(StringDeserializer.class, MyObjectDeserializer.class)` and `companion.produceWithSerializer(StringSerializer.class, MyObjectSerializer.class)`. 
As String is often used as keys, providing  `companion.consumeWithDeserializer(MyObjectDeserializer.class)` with a default `StringDeserializer` for the key and `companion.produceWithSerializer(MyObjectSerializer.class)` with a default `StringSerializer` for the key will be convenient. I can propose a PR for that.

2. When working with AVRO (or other generic serializer / deserializer) the bound of the serializer will be object wich is not very elegant. For example:
```
var recordProduced = new ProducerRecord<Object, Object>("my_topic", keyRecord, valueRecord);
companion.produceWithSerializers(KafkaAvroSerializer.class, KafkaAvroSerializer.class).fromRecords(recordProduced).awaitCompletion();
```
A solution would be to add a type hint to the method:
```
var recordProduced = new ProducerRecord<>("my_topic", keyRecord, valueRecord);
companion.<GenericRecord, GenericRecord>produceWithSerializers(KafkaAvroSerializer.class, KafkaAvroSerializer.class).fromRecords(recordProduced).awaitCompletion();
```
I don't know if something better can be done, we can always add an `as(Class, Class)` method but I'm not sure it's better. Maybe removing the bound to `<K,T>` would do the trick.

3. Consuming no items is not an easy task, this is the piece of code I ended up in my JUnit test : 
```
       Throwable exception = assertThrows(AssertionError.class,
                () -> companion.consumeStrings()
                        .fromTopics("my_topic")
                        .awaitNextRecord(Duration.ofSeconds(2)));
        assertThat(exception.getMessage()).isEqualTo("Expected 1 records in 2000 ms, but only received 0 records.");
```
Maybe we can have a `companion.consumeStrings().fromTopics("my_topic").awaitNoRecords(Duration)` that will throws an exception or return a boolean to false if it receive a list of record (or return the received record). I can propose a PR for that.

4. I need to create a reproducer to show what I means but when you create multiple consume task (multiple `companion.consume()`) on the same test case, the second one receive the messages from the begining, so if we want to have several test methods based on the same topic, we will need to reuse the same consumer for all test methods, due to how it is integrated in Quarkus (via injection), we will need to switch to PER_CLASS lifecycle and things get complicated a lot. I'm not sure if the issue is that each consumer task get it's own group-id or if the offset was not commited so I don't know if there is an easy way to workaround this. I'll comment this discussion with example that will explain more this issue.

**Comment by ozangunalp** [comment]
Hi @loicmathieu,

1. In #1822 added `consumeWithDeserializer` `produceWithSerializer` methods which takes value serde (class or fqcn) as parameter and defaults to string serde for key.

2. I think it makes sense to force the key/value types on the `ProducerBuilder` if the passed serde doesn't provide the type info. 

You can also use the `registerSerde` methods for associating a type with a Serde. You can then pass the object directly to `consume` / `produce` methods. The serde to be used will be resolved from previously registered serdes.
```java
companion.registerSerde(Person.class, new GenericSerializer<>(), new GenericDeserializer<>());

ProducerRecord<String, Person> record = new ProducerRecord<>(topic, new Person("1", 20));
companion.produce(Person.class)
        .fromRecords(record)
        .awaitCompletion();

ConsumerBuilder<String, Person> consumer = companion.consume(Person.class);
ConsumerTask<String, Person> task = consumer.fromTopics(topic, 1).awaitCompletion();
assertThat(task.getRecords()).hasSize(1);
```

Note that you need to call `configure` on your serde if it requires that. There is an example on Quarkus Super heroes : 
https://github.com/quarkusio/quarkus-super-heroes/blob/main/event-statistics/src/test/java/io/quarkus/sample/superheroes/statistics/endpoint/WebSocketsIT.java

3. Again in #1822 Added fluent method awaitNoRecords(Duration) to KafkaTask. It throws assertion error, like other methods, if any records were received in a given duration.

4. This one was also described on #1839. 
Maybe a couple of things are worth noting about the subject:
- Companion API is just a thin/easier-to-use wrapper around Kafka client API. 
- There is very little state kept inside the KafkaCompanion between test cases, I'd say it is safe to reuse it.
- Tests are run against a broker, so produced records or offsets committed to consumer groups are written to that broker. 
- `ConsumerBuilder` and `ProducerBuilder` created via `companion.consume`/`companion.produce` methods are "builders", so templates to create consumers and producers when a produce/consume task is run. You can configure the consumer builder through `with` prefixed methods, for instance, `withGroupId`, `withOffsetReset`, `withAutoCommit` or `withCommitAsyncWhen`. Configuring these would be per use-case.
- It would be useful to delete created topics or delete records from topics before each test case. #1822 adds a `topics().clear ` method to delete all records from all partitions of given topics.
- Please feel free to come up with the reproducer so that we can figure out other use cases.

**Reply by loicmathieu** [reply]
Thanks, as I see the easy ones are already covered by #1822 :)

For point 4 a topic().clear() methods seems a good and easy way to cover this, it is a best practice to clear() any test data in an `@AfterEach` method so it'll be OK for me.
I also can try to use `withGroupId` to see if it can solve the issue.
